README.md
ssldecoder
==========

Check your domain SSL/TLS setup with the SSL Decoder.org API (unofficial)

[![npm](https://img.shields.io/npm/v/ssldecoder.svg?maxAge=3600)](https://github.com/fvdm/nodejs-ssldecoder/blob/master/CHANGELOG.md)
[![Build Status](https://travis-ci.org/fvdm/nodejs-ssldecoder.svg?branch=master)](https://travis-ci.org/fvdm/nodejs-ssldecoder)
[![Dependency Status](https://gemnasium.com/badges/github.com/fvdm/nodejs-ssldecoder.svg)](https://gemnasium.com/github.com/fvdm/nodejs-ssldecoder#runtime-dependencies)
[![Test Coverage](https://codeclimate.com/github/fvdm/nodejs-ssldecoder/badges/coverage.svg)](https://codeclimate.com/github/fvdm/nodejs-ssldecoder/coverage)


* [Node.js](https://nodejs.org)
* [SSL Decoder](https://ssldecoder.org)
* [API documentation](https://github.com/RaymiiOrg/ssl-decoder/blob/master/README.md#json-api)


Example
-------

```js
var ssldecoder = require ('ssldecoder') ();

// Process response data
function processHost (err, data) {
  if (err) {
    console.log (err);
    return;
  }
  
  console.log (data.connection.protocols);
}

// Get certificate details
ssldecoder.host ('myhostname.net', processHost);
```


Installation
------------

`npm install ssldecoder`


#### Configuration

param    | type   | required | default                  | description
:--------|:-------|:---------|:-------------------------|:--------------------------------------
timeout  | number | no       | `20000`                  | Request wait time out in ms, 1000 = 1 sec
endpoint | string | no       | `https://ssldecoder.org` | Send calls to another API


#### Example

```js
// Set timeout to 5 seconds and your own server
var ssldecoder = require ('ssldecoder') ({
  timeout: 5000,
  endpoint: 'https://my-box.tld/path'
});
```


Method .host
------------
**( params, callback )**

Retrieve certificate details from a hostname.


#### Arguments

param    | type             | required | description
:--------|:-----------------|:---------|:-------------------------------------------------------
params   | object or string | yes      | Hostname (string) or [parameters](#parameters) (object)
callback | function         | yes      | `function (err, data) {}`


##### Parameters

param            | type    | required | default                | description
:----------------|:--------|:---------|:-----------------------|:-------------------------------
params.host      | string  | no *     | _resolved from `ip`_   | Hostname
params.ip        | string  | no *     | _resolved from `host`_ | IP address
params.port      | number  | no       | `443`                  | HTTPS port
params.fastcheck | boolean | no       | `false`                | Limited connection data enumeration, no certificate transparency submission


Always provide either `host` or `ip`, or ideally both!
Otherwise the `host` or `ip` is DNS resolved from the other,
selecting only the first result it gets.


#### Example

```js
// Only hostname, let the module lookup the IP
ssldecoder.host ('myhostname.net', processData);

// Only IP, let the module lookup the hostname
var params = {
  ip: '2a01:7c8:aab3:35a::3',
  fastcheck: true
};

ssldecoder.host (params, processData);

// Full control with hostname and IP
var params = {
  host: 'myhostname.net',
  ip: '1.2.3.4'
};

ssldecoder.host (params, processData);
```


Method .csr
-----------
**( csr, callback )**

Process CSR PEM


#### Arguments

argument | type     | required | description
:--------|:---------|:---------|:--------------------------------------------
csr      | string   | yes      | Either `/path/to/csr.pem` or full PEM string
callback | function | yes      | `function (err, data) {}`


#### Example

```js
// Read PEM from path
ssldecoder.csr ('csr.pem', processCSR);

// Send PEM text
var pem = '-----BEGIN CERTIFICATE REQUEST-----\n...';

ssldecoder.csr (pem, processCSR);
```


Unlicense
---------

This is free and unencumbered software released into the public domain.

Anyone is free to copy, modify, publish, use, compile, sell, or
distribute this software, either in source code form or as a compiled
binary, for any purpose, commercial or non-commercial, and by any
means.

In jurisdictions that recognize copyright laws, the author or authors
of this software dedicate any and all copyright interest in the
software to the public domain. We make this dedication for the benefit
of the public at large and to the detriment of our heirs and
successors. We intend this dedication to be an overt act of
relinquishment in perpetuity of all present and future rights to this
software under copyright law.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

For more information, please refer to <http://unlicense.org>


Author
------

[Franklin van de Meent](https://frankl.in)


CHANGELOG.md
#### 1.0.1 (2016-6-3)

##### Chores

* **package:** Update dependencies versions ([66004d45](https://github.com/fvdm/nodejs-ssldecoder/commit/66004d45dc0a4e5b56c701913b547fb41fe8f772))

##### Documentation Changes

* **badges:**
  * Add CodeClimate coverage ([f6036877](https://github.com/fvdm/nodejs-ssldecoder/commit/f6036877c8fc03f70f97be7870eb5c0dcfec748d))
  * Add npm version for changelog ([abc76ba7](https://github.com/fvdm/nodejs-ssldecoder/commit/abc76ba74c0bab8296d0ec7eb2c6161dcf3a7033))
  * Deeplink Gemnasium to dependencies tab ([fd21c540](https://github.com/fvdm/nodejs-ssldecoder/commit/fd21c5403f039f9c5413791761fce84ab5474efb))

##### Other Changes

* **undefined:**
  * add node v6 to Travis config ([eb9e2db6](https://github.com/fvdm/nodejs-ssldecoder/commit/eb9e2db6c4bf297d0b3fea76c216f47825b5cc6f))
  * added Unlicense text ([56b43d23](https://github.com/fvdm/nodejs-ssldecoder/commit/56b43d235f9168d05aae0483e9dcd02dc036628c))
  * added dependencies badge ([51bd4134](https://github.com/fvdm/nodejs-ssldecoder/commit/51bd41349fdbff202881a7a96fdc254c77790262))

##### Refactors

* **package:**
  * Add CodeClimate and Istanbul devDependencies ([d0355c4d](https://github.com/fvdm/nodejs-ssldecoder/commit/d0355c4d21677acb6d3adafdfaa783d63242d4a5))
  * Run npm test with test.sh ([b0283e6f](https://github.com/fvdm/nodejs-ssldecoder/commit/b0283e6fd327a539fcc8046c0f1e6643ea028db3))
* **config:**
  * ESlint warn on console.logs ([8ddea09c](https://github.com/fvdm/nodejs-ssldecoder/commit/8ddea09ce4a17f64cfabf84ba2872fa3e2e6db34))
  * Add .*_history to .gitignore ([73503ba5](https://github.com/fvdm/nodejs-ssldecoder/commit/73503ba5320212b5d5e08e88319a81435175ac17))

##### Tests

* **config:**
  * Revert Travis to simplicity ([9fab4e89](https://github.com/fvdm/nodejs-ssldecoder/commit/9fab4e8981e8e43af920ac2acb5ae3ef26513be0))
  * Minimal Travis environment ([92687df9](https://github.com/fvdm/nodejs-ssldecoder/commit/92687df92f42ec078c618fdc030cc350fd15b728))
  * Seperated jobs for each test ([d9c17504](https://github.com/fvdm/nodejs-ssldecoder/commit/d9c17504de0c9e6322938665fa2476255015dc2d))
  * Add CodeClimate config ([2c8fe33a](https://github.com/fvdm/nodejs-ssldecoder/commit/2c8fe33a2174ddddb9771a98d976121382b7d110))
* **cleanup:** Rewrite script to simplicity ([950f2a03](https://github.com/fvdm/nodejs-ssldecoder/commit/950f2a037a5778c06d30a7badedc0dc3d110c0e8))
* **refactor:** Add bash script for running tests ([40fbf41d](https://github.com/fvdm/nodejs-ssldecoder/commit/40fbf41df20b44c29ea5a9bca194f1fb23ef95c3))



test.js
var dotest = require ('dotest');
var app = require ('./');

var config = {
  timeout: process.env.Timeout || 20000,
  endpoint: process.env.Endpoint || null
};

var ssldecoder = app (config);

var csr = [];
var pub = [];

csr.push ('-----BEGIN CERTIFICATE REQUEST-----');
csr.push ('MIIC6zCCAdMCAQAwgaUxCzAJBgNVBAYTAk5MMRYwFAYDVQQIEw1FeGFtcGxlLVN0');
csr.push ('YXRlMRUwEwYDVQQHEwxFeGFtcGxlIENpdHkxFTATBgNVBAoTDFRoZSBJbnRlcm5l');
csr.push ('dDEdMBsGA1UECxMUR2xvYmFsIGNvbW11bmljYXRpb24xEjAQBgNVBAMTCWxvY2Fs');
csr.push ('aG9zdDEdMBsGCSqGSIb3DQEJARYOaW5mb0Bsb2NhbGhvc3QwggEiMA0GCSqGSIb3');
csr.push ('DQEBAQUAA4IBDwAwggEKAoIBAQC90DnuvyjUjsIFYVrMR+QO6fb/vUNH2E/lsxbw');
csr.push ('r/RMK6vASm9RRDpTeuuFt+cK+M10n2qdt+nSqzNDaNV3uBS7rStyq5XFy7Js2Gjj');
csr.push ('//+JWuHH3KyqL9tlD4QLP4FBVUidahG6RDtVIAWCf0kTYjYCe4AHnmmvLiigomgs');
csr.push ('l4455BwGpiiqHv+EW1Y5K3kPDojIFayRZunJl3ZoaPkHHBjWYL2fHnF81oyMOoDA');
csr.push ('omSC17kDysN7OdWt4SaYrAUSwePqRlwNRE5Xc1rpndt+SgNOqVnEhsFUhd4GrHf9');
csr.push ('kFfaC0DAwOI6HUkjmo3v86ty0ZEoYcWQPbU8X9YVwkf9GW7VAgMBAAGgADANBgkq');
csr.push ('hkiG9w0BAQUFAAOCAQEAGbAbjZ8eaLTnZNlvEcexMQPXf2fGW1w0No2JjDzwiLQn');
csr.push ('vxuCCqqu8jgU3txkodV16H+Yd3xHwIdFCiZJTaNptSty6vXROREvRrKKB4sHqfWF');
csr.push ('ZSaaugtwGpU9+2LY5/i+rmgO6iiaOsvWx/pCiDPkazer81k1YpneGmRmLwuzLQsm');
csr.push ('NYFYRH+ijk3A3kcrlMZnVVskkpLDMEchfSkz7M1W7MYHLwmiclP9tOybr+z2xW7G');
csr.push ('zt8H5ErnV1AcVFTcr/wKjJyZ1SxAFZvSEs0u5pcmXcqJQKCzd1sFghGXO9lC3RBX');
csr.push ('WkiNl2pKvnoNR7EnqJqgntD+4Ysncr1QcY+CQrKjtQ==');
csr.push ('-----END CERTIFICATE REQUEST-----');

csr = csr.join ('\n');

pub.push ('-----BEGIN PUBLIC KEY-----');
pub.push ('MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAvdA57r8o1I7CBWFazEfk');
pub.push ('Dun2/71DR9hP5bMW8K/0TCurwEpvUUQ6U3rrhbfnCvjNdJ9qnbfp0qszQ2jVd7gU');
pub.push ('u60rcquVxcuybNho4///iVrhx9ysqi/bZQ+ECz+BQVVInWoRukQ7VSAFgn9JE2I2');
pub.push ('AnuAB55pry4ooKJoLJeOOeQcBqYoqh7/hFtWOSt5Dw6IyBWskWbpyZd2aGj5BxwY');
pub.push ('1mC9nx5xfNaMjDqAwKJkgte5A8rDeznVreEmmKwFEsHj6kZcDUROV3Na6Z3bfkoD');
pub.push ('TqlZxIbBVIXeBqx3/ZBX2gtAwMDiOh1JI5qN7/OrctGRKGHFkD21PF/WFcJH/Rlu');
pub.push ('1QIDAQAB');
pub.push ('-----END PUBLIC KEY-----');
pub.push ('');

pub = pub.join ('\n');


dotest.add ('Module', function (test) {
  test ()
    .info ('config.timeout:   ' + config.timeout)
    .info ('config.endpoint:  ' + (config.endpoint ? 'custom' : 'live'))
    .isFunction ('fail', 'exports', app)
    .isObject ('fail', 'interface', ssldecoder)
    .isFunction ('fail', '.host', ssldecoder && ssldecoder.host)
    .isFunction ('fail', '.csr', ssldecoder && ssldecoder.csr)
    .done ();
});


dotest.add ('Method .csr - from text', function (test) {
  ssldecoder.csr (csr, function (err, data) {
    test (err)
      .isObject ('fail', 'data', data)
      .isArray ('fail', 'data.chain', data && data.chain)
      .isNotEmpty ('fail', 'data.chain', data && data.chain)
      .isObject ('fail', 'data.chain[0]', data && data.chain && data.chain [0])
      .isExactly ('warn', 'data.chain[0].key', data && data.chain && data.chain [0] .key, pub)
      .done ();
  });
});


dotest.add ('Method .csr - from file', function (test) {
  ssldecoder.csr ('./test.csr', function (err, data) {
    test (err)
      .isObject ('fail', 'data', data)
      .isArray ('fail', 'data.chain', data && data.chain)
      .isNotEmpty ('fail', 'data.chain', data && data.chain)
      .isObject ('fail', 'data.chain[0]', data && data.chain && data.chain [0])
      .isExactly ('warn', 'data.chain[0].key', data && data.chain && data.chain [0] .key, pub)
      .done ();
  });
});


dotest.add ('Method .host', function (test) {
  ssldecoder.host ('myhostname.net', function (err, data) {
    var chain = data && data.chain;
    var chainOcsp = chain && chain [0] && chain [0] .ocsp;
    var connection = data && data.connection;
    var connChain = connection && connection.chain;
    var protocols = connection && connection.protocols;
    var ciphers = connection && connection.supported_ciphersuites;

    test (err)
      .isObject ('fail', 'data', data)
      .isObject ('fail', 'data.connection', connection)
      .isExactly ('fail', 'data.connection.checked_hostname', connection && connection.checked_hostname, 'myhostname.net')
      .isArray ('fail', 'data.connection.chain', connChain)
      .isNotEmpty ('fail', 'data.connection.chain', connChain)
      .isObject ('fail', 'data.connection.chain[0]', connChain && connChain [0])
      .isObject ('fail', 'data.connection.protocols', protocols)
      .isExactly ('warn', 'data.connection.protocols [tlsv1.2]', protocols ['tlsv1.2'], true)
      .isExactly ('warn', 'data.connection.protocols [sslv2]', protocols.sslv2, false)
      .isArray ('fail', 'data.connection.supported_ciphersuites', ciphers)
      .isString ('fail', 'data.connection.supported_ciphersuites[0]', ciphers [0])
      .isArray ('fail', 'data.chain', chain)
      .isNotEmpty ('fail', 'data.chain', chain)
      .isObject ('fail', 'data.chain[0]', chain && chain [0])
      .isObject ('fail', 'data.chain[0].cert_data', chain && chain [0] && chain [0] .cert_data)
      .isArray ('fail', 'data.chain[0].ocsp', chainOcsp)
      .isObject ('warn', 'data.chain[0].ocsp[0]', chainOcsp && chainOcsp [0])
      .done ();
  });
});


dotest.run ();


test.sh
#!/bin/bash
result=0

eslint *.js || result=1

if [ "$TRAVIS_BRANCH" == "master" ] && [ "$CODECLIMATE_REPO_TOKEN" != "" ]; then
  istanbul cover test.js --print none --report lcovonly || result=1
  [ "$result" -eq "0" ] && codeclimate-test-reporter < ./coverage/lcov.info || result=1
else
  node test.js || result=1
fi

exit $result



index.js
/*
Name:            ssldecoder
Description:     Access SSLDecoder.org API methods (unofficial)
Author:          Franklin van de Meent (https://frankl.in)
License:         Unlicense (Public Domain, see LICENSE file)
Source code:     https://github.com/fvdm/nodejs-ssldecoder
Contact & bugs:  https://github.com/fvdm/nodejs-ssldecoder/issues
*/


var httpreq = require ('httpreq');
var dns = require ('dns');
var fs = require ('fs');

var config = {
  endpoint: 'https://ssldecoder.org',
  timeout: 20000
};


/**
 * Convert an object to an array
 *
 * @param obj {object} - The object to convert
 * @returns {array} - The array
 */

function Object2Array (obj) {
  var key;
  var result = [];

  if (obj instanceof Object) {
    for (key in obj) {
      result.push (obj [key]);
    }

    return result;
  }

  return obj;
}


/**
 * Deep object iteration
 * Loosely based on http://stackoverflow.com/a/25334210
 *
 * @param obj {object} - The object to process
 * @param func {function} - `function (val, key) { return val + 2; }`
 * @returns {object} - The processed object
 */

function deepMap (obj, func) {
  return Object.keys (obj) .reduce (function (acc, k) {
    if (typeof obj [k] === 'object') {
      acc [k] = deepMap (obj [k], func);
    } else {
      acc [k] = func (obj [k], k);
    }

    return acc;
  }, {});
}


/**
 * Look for certain keys and convert their value to a boolean
 *
 * '1'     -> true
 * ''      -> false
 * 'true'  -> true
 * 'false' -> false
 *
 * @param obj {object} - The object to process
 * @param keys {array} - Keys to look for
 * @returns {object} - The processed object
 */

function boolDozer (obj, keys) {
  obj = deepMap (obj, function (val, key) {
    if (!!~keys.indexOf (key)) {
      if (typeof val === 'string' && !!val.match (/^(1|true)$/)) {
        return true;
      }

      if (typeof val === 'string' && !!val.match (/^(|false)$/)) {
        return false;
      }

      return val;
    }

    return val;
  });

  return obj;
}


/**
 * Process API response
 *
 * @callback callback
 * @param err {Error, null} - Error
 * @param res {object} - Response object
 * @param callback {function} - `function (err, data) {}`
 * @returns {void}
 */

function processResponse (err, res, callback) {
  var data = res && res.body || '';
  var error = null;

  if (err) {
    error = new Error ('Client error');
    error.error = err;
    callback (error);
    return;
  }

  try {
    data = JSON.parse (data);
  } catch (e) {
    error = new Error ('Invalid response');
    error.statusCode = res.statusCode;
    error.error = e;
    callback (error);
    return;
  }

  if (data && data.error) {
    error = new Error ('API error');
    error.statusCode = res.statusCode;
    error.error = data.error;
    callback (error);
    return;
  }

  if (data && !data.data) {
    error = new Error ('Invalid response');
    error.statusCode = res.statusCode;
    error.data = data;
    callback (error);
    return;
  }

  callback (null, data.data);
}


/**
 * Send API request
 *
 * @callback callback
 * @param params {object} - Data fields to send
 * @param callback {function} - `function (err, data) {}`
 * @returns {void}
 */

function sendRequest (params, callback) {
  var options = {
    url: config.endpoint + '/json.php',
    parameters: params,
    method: 'GET',
    timeout: config.timeout,
    headers: {
      'User-Agent': 'ssldecoder (https://github.com/fvdm/nodejs-ssldecoder)'
    }
  };

  httpreq.doRequest (options, function (err, res) {
    processResponse (err, res, callback);
  });
}


/**
 * Process host response data
 *
 * @callback callback
 * @param err {Error, null} - Response error
 * @param data {object} - Response data
 * @param callback {function} - `function (err, data) {}`
 */

function processHost (err, data, callback) {
  var itm;
  var i;

  var bools = [
    'tlsv1.2',
    'tlsv1.1',
    'tlsv1.0',
    'sslv3',
    'sslv2',
    'working',
    'heartbeat',
    'ca',
    'general',
    'cert_issued_in_future',
    'cert_expired',
    'cert_expires_in_less_than_thirty_days',
    'issuer_valid'
  ];

  if (err) {
    callback (err);
    return;
  }

  // Set boolean values
  data = boolDozer (data, bools);

  // Convert numbered objects to arrays
  data.connection.chain = Object2Array (data.connection.chain);
  data.connection.supported_ciphersuites = Object2Array (data.connection.supported_ciphersuites);
  data.chain = Object2Array (data.chain);

  for (i = 0; i < data.chain.length; i++) {
    itm = data.chain [i];

    if (itm.ocsp) {
      itm.ocsp = Object2Array (itm.ocsp);
    }

    if (itm.crl) {
      itm.crl = Object2Array (itm.crl);
    }

    data.chain [i] = itm;
  }

  // All done
  callback (null, data);
}


/**
 * Method .host
 *
 * @callback callback
 * @param params {object} - Method parameters
 * @param callback {function} - `function (err, data) {}`
 * @returns {void}
 */

function methodHost (params, callback) {
  var send = {
    host: '',
    port: 443,
    fastcheck: 0
  };

  // "host"
  if (typeof params === 'string') {
    params = {
      host: params
    };
  }

  // set parameters
  send.port = params.port || send.port;
  send.fastcheck = params.fastcheck || send.fastcheck;

  // host:ip
  if (typeof params.host === 'string' && !!params.host.match (/:/)) {
    send.host = params.host;
    sendRequest (send, function (rErr, rData) {
      processHost (rErr, rData, callback);
    });
    return;
  }

  // host && ip
  if (params.host && params.ip) {
    send.host = params.host + ':' + params.ip;
    sendRequest (send, function (rErr, rData) {
      processHost (rErr, rData, callback);
    });
    return;
  }

  // host && !ip
  if (params.host && !params.ip) {
    dns.resolve (params.host, function (err, res) {
      if (err) {
        callback (err);
        return;
      }

      send.host = params.host + ':' + res [0];
      sendRequest (send, function (rErr, rData) {
        processHost (rErr, rData, callback);
      });
    });

    return;
  }

  // !host && ip
  if (!params.host && params.ip) {
    dns.reverse (params.ip, function (err, res) {
      if (err) {
        callback (err);
        return;
      }

      send.host = res [0] + ':' + params.ip;
      sendRequest (send, function (rErr, rData) {
        processHost (rErr, rData, callback);
      });
    });

    return;
  }

  callback (new Error ('Invalid host or IP address'));
}


/**
 * Method .csr
 *
 * @callback callback
 * @param csr {string} - PEM string or filepath
 * @param callback {function} - `function (err, data) {}`
 * @returns {void}
 */

function methodCsr (csr, callback) {
  var params = {
    csr: csr
  };

  if (!csr || typeof csr !== 'string') {
    callback (new Error ('CSR must be a string'));
    return;
  }

  if (!!csr.match (/^-----BEGIN CERTIFICATE/)) {
    sendRequest (params, function (err, data) {
      if (err) {
        callback (err);
        return;
      }

      data.chain = Object2Array (data.chain);
      callback (null, data);
    });

    return;
  }

  fs.readFile (csr, { encoding: 'utf8' }, function (fsErr, file) {
    if (fsErr) {
      callback (fsErr);
      return;
    }

    params.csr = file;

    sendRequest (params, function (err, data) {
      if (err) {
        callback (err);
        return;
      }

      data.chain = Object2Array (data.chain);
      callback (null, data);
    });
  });
}


/**
 * Configuration
 *
 * @param [cnf] {object} - Config object
 * @param [cnf.timeout = 5000] {number} - Time out in ms
 * @param [cnf.endpoint = https://ssldecoder.org] {string} - API endpoint
 * @returns {object} - Interface methods
 */

function setup (cnf) {
  if (cnf instanceof Object) {
    config.timeout = cnf.timeout || config.timeout;
    config.endpoint = cnf.endpoint || config.endpoint;
  }

  return {
    host: methodHost,
    csr: methodCsr
  };
}

module.exports = setup;
